node/src/tasks/user_submitted_file.rs

```diff
@@ -0,0 +1,111 @@
+use crate::services::file_transfer::commands::FileTransferServiceInterface;
+use crate::tasks::AcceptedBspVolunteer;
+use crate::tasks::StorageHubHandler;
+use crate::tasks::StorageHubHandlerConfig;
+use file_manager::traits::FileStorage;
+use log::{error, info};
+use sc_network::Multiaddr;
+use sc_network::PeerId;
+use sp_core::Blake2Hasher;
+use sp_core::Hasher;
+use std::str::FromStr;
+use storage_hub_infra::event_bus::EventHandler;
+use storage_hub_infra::types::Metadata;
+
+const LOG_TARGET: &str = "user-submitted-file-task";
+
+/// Handles the events related to users submitting files to be stored.
+pub struct UserSubmittedFileTask<SHC: StorageHubHandlerConfig> {
+    storage_hub_handler: StorageHubHandler<SHC>,
+}
+
+impl<SHC: StorageHubHandlerConfig> Clone for UserSubmittedFileTask<SHC> {
+    fn clone(&self) -> Self {
+        Self {
+            storage_hub_handler: self.storage_hub_handler.clone(),
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> UserSubmittedFileTask<SHC> {
+    pub fn new(storage_hub_handler: StorageHubHandler<SHC>) -> Self {
+        Self {
+            storage_hub_handler,
+        }
+    }
+}
+
+impl<SHC: StorageHubHandlerConfig> EventHandler<AcceptedBspVolunteer>
+    for UserSubmittedFileTask<SHC>
+{
+    async fn handle_event(&self, event: AcceptedBspVolunteer) -> anyhow::Result<()> {
+        info!(
+            target: LOG_TARGET,
+            "Handling file submitted by user {:?} to BSP with location {:?}",
+            event.owner,
+            event.location,
+        );
+
+        // TODO: use `Multiaddr` instead of `String`.
+        let multiaddresses = event.multiaddresses;
+        let file_metadata = Metadata {
+            owner: event.owner.to_string(),
+            size: event.size.into(),
+            // TODO: use `FileLocation` instead of `String`.
+            location: format!("{:?}", event.location.into_inner()),
+            fingerprint: event.fingerprint,
+        };
+        let chunk_count = file_metadata.chunk_count();
+
+        // TODO(Arthur): double check this, I'm assuming Blake2 as the Trie hash function.
+        let file_key = Blake2Hasher::hash(&serde_json::to_vec(&file_metadata)?.to_owned());
+
+        let peer_ids = multiaddresses
+            .iter()
+            .map(|multiaddr| {
+                let multiaddr_str = multiaddr
+                    .split("/")
+                    .last()
+                    .expect("Multiaddress of Bsp should not be empty.");
+                let multiaddr =
+                    Multiaddr::from_str(multiaddr_str).expect("Failed to convert into Multiaddr");
```

Comment:
Probably we would like to continue with any BSPs we can instead of failing everything? Or is the assumption that this is verified by the runtime before emitting the event too?

Reply:
I wouldn’t fail here. Instead continue until the end and if there are no `peer_ids` at the end of this, I would fallback to querying the storage providers for their multiaddresses, and find any that might have been added after they volunteered. It is highly unlikely that there will be a change, but it could happen.

Reply:
You are right, I removed the failures and now just proceed with getting whatever PeerIds are available.